primefield/src/monty.rs

```diff
@@ -0,0 +1,178 @@
+///     "P-256 field modulus"
+/// );
+/// ```
+#[macro_export]
+macro_rules! monty_field_params {
+    ($name:ident, $modulus_hex:expr, $uint_type:ty, $byte_size:ty, $byte_order:expr, $doc:expr) => {
+        $crate::bigint::const_monty_params!($name, $uint_type, $modulus_hex, $doc);
+
+        impl $crate::MontyFieldParams<{ <$uint_type>::LIMBS }> for $name {
+            type ByteSize = $byte_size;
+            const BYTE_ORDER: $crate::ByteOrder = $byte_order;
+        }
+    };
+}
+
+/// Field element type which uses an internal Montgomery form representation.
+#[derive(Clone, Copy)]
+pub struct MontyFieldElement<MOD: MontyFieldParams<LIMBS>, const LIMBS: usize>(
+    MontyForm<MOD, LIMBS>,
+);
+
+/// Extension trait for defining additional field parameters beyond the ones provided by
+/// [`ConstMontyParams`].
+pub trait MontyFieldParams<const LIMBS: usize>: ConstMontyParams<LIMBS> {
+    /// Size of a field element when serialized as bytes.
+    type ByteSize: ArraySize;
+
+    /// Byte order to use when serializing a field element as byte.
+    const BYTE_ORDER: ByteOrder;
+}
+
+impl<MOD: MontyFieldParams<LIMBS>, const LIMBS: usize> MontyFieldElement<MOD, LIMBS> {
+    /// Zero element (additive identity).
+    pub const ZERO: Self = Self(MontyForm::ZERO);
+
+    /// Multiplicative identity.
+    pub const ONE: Self = Self(MontyForm::ONE);
+
+    /// Number of limbs used by the internal integer representation.
+    pub const LIMBS: usize = LIMBS;
+
+    /// Decode field element from a canonical bytestring representation.
+    #[inline]
+    pub fn from_bytes(repr: &Array<u8, MOD::ByteSize>) -> CtOption<Self>
+    where
+        Uint<LIMBS>: ArrayEncoding,
+    {
+        debug_assert!(repr.len() <= MOD::ByteSize::USIZE);
+        let mut byte_array = ByteArray::<Uint<LIMBS>>::default();
+        let offset = MOD::ByteSize::USIZE.saturating_sub(repr.len());
+
+        let uint = match MOD::BYTE_ORDER {
+            ByteOrder::BigEndian => {
+                byte_array[offset..].copy_from_slice(repr);
+                Uint::from_be_byte_array(byte_array)
+            }
+            ByteOrder::LittleEndian => {
+                byte_array[..offset].copy_from_slice(repr);
+                Uint::from_le_byte_array(byte_array)
+            }
+        };
+
+        Self::from_uint(&uint)
+    }
+
+    /// Decode field element from a canonical byte slice.
+    ///
+    /// Slice is expected to be zero padded to the expected byte size.
+    #[inline]
+    pub fn from_slice(slice: &[u8]) -> Option<Self>
+    where
+        Uint<LIMBS>: ArrayEncoding,
+    {
+        let array = Array::try_from(slice).ok()?;
+        Self::from_bytes(&array).into()
+    }
+
+    /// Decode a field element from hex-encoded bytes.
+    ///
+    /// Does *not* perform a check that the field element does not overflow the order.
+    ///
+    /// This method is primarily intended for defining internal constants.
+    pub const fn from_hex(hex: &str) -> Self {
```

Comment:
One drawback is this is no longer `pub(crate)`.

I tried adding compile-time checks the modulus is in range, but it seems the relevant methods to do `const fn` comparisons are currently `pub(crate)` (i.e. `Uint::lt`)

Reply:
I added the checks which makes it slightly less worrisome